fireplace-swing/src/main/java/com/github/bric3/fireplace/flamegraph/FlameGraphPainter.java

```diff
@@ -297,47 +292,39 @@ private void paintNodeFrameRectangle(Graphics2D g2, Rectangle frameRect, T node,
                        frameRectSurface.width - textBorder * 2 - frameBorderWidth * 2,
                        text -> {
                            g2.setColor(Colors.foregroundColor(bgColor));
-                           g2.drawString(text, textBorder + frameBorderWidth, getFrameBoxTextOffset(g2));
+                           g2.drawString(text, frameRect.x + textBorder, frameRect.y + getFrameBoxTextOffset(g2));
                        });
-        g2.dispose();
     }
 
-    private void paintRootFrameRectangle(Graphics2D g2, Rectangle rect, String str, Color bgColor, boolean minimapMode) {
-        var frameRectSurface = paintFrameRectangle(g2, rect, bgColor, frameBorderColor, minimapMode);
+    private void paintRootFrameRectangle(Graphics2D g2, Rectangle rect, String str, Color bgColor, Color gapColor, boolean minimapMode) {
+        var frameRectSurface = paintFrameRectangle(g2, rect, bgColor, gapColor, minimapMode);
         if (minimapMode) {
             return;
         }
         paintFrameText(str,
                        g2,
-                       frameRectSurface.width - textBorder * 2 - frameBorderWidth * 2,
+                       frameRectSurface.width - textBorder * 2 - frameGapWidth * 2,
                        text -> {
                            g2.setColor(Colors.foregroundColor(bgColor));
                            g2.drawString(text, textBorder + frameBorderWidth, getFrameBoxTextOffset(g2));
                        });
-        g2.dispose();
     }
 
-    private Rectangle2D.Double paintFrameRectangle(Graphics2D g2, Rectangle frameRect, Color bgColor, Color frameBorderColor, boolean minimapMode) {
+    private Rectangle paintFrameRectangle(Graphics2D g2, Rectangle frameRect, Color bgColor, Color frameGapColor, boolean minimapMode) {
         var borderWidth = minimapMode ?
                           0 :
-                          paintFrameBorder ? frameBorderWidth : 0;
-
-        if (paintFrameBorder && !minimapMode) {
-            var growBoxSurface = new Rectangle.Double(frameRect.x,
-                                                      frameRect.y,
-                                                      frameRect.width + borderWidth,
-                                                      frameRect.height + borderWidth);
-            g2.setColor(frameBorderColor);
-            g2.fill(growBoxSurface);
+                          frameGapEnabled ? frameGapWidth : 0;
+
+        if (frameGapEnabled && !minimapMode) {
+            g2.setColor(frameGapColor);
+            g2.fill(frameRect);
```

Comment:
I had a lot of trouble to get this right. I would welcome something way simpler.

At this time there still an issue when the mouse is moving over a frame.

This pr
<img width="571" alt="image" src="https://user-images.githubusercontent.com/803621/158417066-640b7350-1f91-4d97-a778-b2727ef7ab75.png">

vs master
<img width="575" alt="image" src="https://user-images.githubusercontent.com/803621/158417970-ed2ff317-925e-4790-9c06-930e9abe2f89.png">


Thinking about it I thing there's actually two concepts, the gap between frames, and the border when hovering a frame.

Reply:
On the clipping topic, what do you think about set the clipping for a frame to `frameRect`?

Reply:
> On the clipping topic, what do you think about set the clipping for a frame to `frameRect`?

Definitely can be done, although there isn't anything being drawn that *should* require clipping, so this would just be a safety net (with a small performance penalty).

Reply:
I pushed a change that should give you a better result on the hover border - try it and let me know.

Reply:
Follow-up: https://github.com/bric3/fireplace/pull/27#discussion_r827902076